src/main/scala/chisel3/util/Reg.scala

```diff
@@ -123,6 +123,91 @@ object ShiftRegister {
     compileOptions:      CompileOptions
   ): T =
     ShiftRegisters(in, n, resetData, en).lastOption.getOrElse(in)
+
+  /** Returns the n-cycle delayed version of the input signal (SyncReadMem-based ShiftRegister implementation).
+    *
+    * @param in input to delay
+    * @param n  number of cycles to delay
+    * @param en enable the shift
+    * @param use_sp_mem single port or dual port SRAM based implementation
+    * @param name name of SyncReadMem object
+    */
+  def apply[T <: Data](in: T, n: Int, en: Bool, use_sp_mem: Boolean, name: String): T =
```

Comment:
How about using `name: Option[String]`? `null` in Scala is a bad design choice.

Reply:
Actually the whole idea is summarized in a [recent issue](https://github.com/chipsalliance/chisel3/issues/2889). In other words, please have a look at [slide 34](http://pages.hmc.edu/harris/class/e158/lect15-sram.pdf): for larger (i.e., wider and deeper) shift registers it is much more power and area efficient to implement them using SRAM than with flip-flops. If dual-port SRAM is available (i.e., if such memory compiler exists) that is a straightforward path. If you want to be even more power and area efficient, or if dual-port memory compiler isn't available in the target technology, you can alternatively use two half-size single-port SRAMs in ping-pong configuration. This is well explained in, say, [Chapter 9](https://onlinelibrary.wiley.com/doi/epdf/10.1002/9780470822500.ch9) (pp. 208-211) of the textbook [OFDM Baseband Receiver Design for Wireless Communications](https://onlinelibrary.wiley.com/doi/book/10.1002/9780470822500) by [Tzi-Dar Chiueh](http://cc.ee.ntu.edu.tw/~chiueh) and [Pei-Yun Tsai](https://scholars.ncu.edu.tw/en/persons/pei-yun-tsai). Even though it is certainly possible to have any number of banks, just as you've said, there is no real gain in power and area for anything larger than two due to (control) logic overhead.

By the way, thanks on the suggestion how to eliminate `null`. You can review it again.

Reply:
Thanks for the clear explanation! Would you mind renaming the arg to `useDualSRAMPort` and default to `false` or something similar and provide a documentation to which? That will be much clear to users.